--- a/trunk/build_firmware_modify
+++ b/trunk/build_firmware_modify
@@ -499,6 +499,9 @@ if [ "$CONFIG_RALINK_MT7620" = "y" ] ; then
 fi
 if [ "$CONFIG_RALINK_MT7621" = "y" ] ; then
+	if [ "$CONFIG_FIRMWARE_CPU_1000MHZ" = "y" ] ; then
+		func_enable_kernel_param "CONFIG_RALINK_MT7621_PLL1000"
+	fi
 	if [ "$CONFIG_FIRMWARE_CPU_900MHZ" = "y" ] ; then
 		func_enable_kernel_param "CONFIG_RALINK_MT7621_PLL900"
 	fi
--- a/trunk/linux-3.4.x/arch/mips/rt2880/Kconfig
+++ b/trunk/linux-3.4.x/arch/mips/rt2880/Kconfig
@@ -107,6 +107,11 @@ config  RALINK_MT7620_PLL600
 	depends on (RALINK_MT7620)
 	default n
 
+config  RALINK_MT7621_PLL1000
+	bool "Set MT7621 CPU clock to 1000MHz (Override Uboot config)"
+	depends on (RALINK_MT7621)
+	default n
+
 config  RALINK_MT7621_PLL900
 	bool "Set MT7621 CPU clock to 900MHz (Override Uboot config)"
 	depends on (RALINK_MT7621)
--- a/trunk/linux-3.4.x/arch/mips/rt2880/init.c
+++ b/trunk/linux-3.4.x/arch/mips/rt2880/init.c
@@ -354,7 +354,16 @@ static void prom_init_sysclk(void)
 		break;
 	case 1: /* CPU PLL */
 		reg = (*(volatile u32 *)(RALINK_MEMCTRL_BASE + 0x648));
-#if defined(CONFIG_RALINK_MT7621_PLL900)
+#if defined(CONFIG_RALINK_MT7621_PLL1000)
+		{
+			u32 fdiv_val = (xtal == 25) ? 39 : 49;
+			reg &= ~(0x7F0 | 0x0F);
+			reg |= ((fdiv_val << 4) | 0x02);
+			(*((volatile u32 *)(RALINK_MEMCTRL_BASE + 0x648))) = reg;
+			udelay(10);
+		}
+#elif defined(CONFIG_RALINK_MT7621_PLL900)
 		if ((reg & 0xff) != 0xc2) {
 			reg &= ~(0xff);
 			reg |=  (0xc2);
